### ISSUE TITLE
🐛 Bug Report: Refreshing OAuth token... forever and stop working

### ISSUE THREAD

**Issue by freedit-dev** [redlib-org/redlib#150]
It occurs several times that my instance stop working and the last log is always

`INFO  redlib::oauth > [⌛] 86279s Elapsed! Refreshing OAuth token...`

And restarting redlib always fixed it.

Maybe there should be timeout for refreshing token.

Thanks for your great work.

**Comment by sigaloid** [comment]
Weird. The refresh code is literally just logging in again, which is what happens at launch.

https://github.com/redlib-org/redlib/blob/6c2579cda9efaeea0bfd8e4a53b298b2b1ef9c6e/src/oauth.rs#L102-L108

I suppose it could have something to do with the header cache not being cleared.

Can you reproduce this consistently? If so, can you try running off of this branch? https://github.com/redlib-org/redlib/tree/try_fix_oauth_refresh

**Comment by freedit-dev** [comment]
The latest commit(fix_oauth_ratelimit)

The instance stops responding(no error log) and the last log:

```log
Jun 27 01:41:19 base redlib[4038062]:  ERROR redlib::utils  > Error page rendered: Reddit error 400 "null": "Bad Request"
Jun 27 01:41:57 base redlib[4038062]:  ERROR redlib::utils  > Error page rendered: Reddit error 400 "null": "Bad Request"
Jun 27 01:43:33 base redlib[4038062]:  WARN  redlib::client > Rate limit 9 is low. Spawning force_refresh_token()
Jun 27 01:43:33 base redlib[4038062]:  INFO  redlib::oauth  > [✅] Success - Retrieved token "eyJhbGciOiJSUzI1NiIsImtpZCI6IlNI...", expires in 86399
Jun 27 01:43:33 base redlib[4038062]:  INFO  redlib::oauth  > Refreshing OAuth token... success
Jun 27 01:43:33 base redlib[4038062]:  WARN  redlib::client > Rate limit 8 is low. Spawning force_refresh_token()
Jun 27 01:43:33 base redlib[4038062]:  INFO  redlib::oauth  > [✅] Success - Retrieved token "eyJhbGciOiJSUzI1NiIsImtpZCI6IlNI...", expires in 86399
Jun 27 01:43:33 base redlib[4038062]:  INFO  redlib::oauth  > Refreshing OAuth token... success
Jun 27 01:47:35 base redlib[4038062]:  WARN  redlib::client > Rate limit 9 is low. Spawning force_refresh_token()
Jun 27 01:47:35 base redlib[4038062]:  WARN  redlib::client > Rate limit 8 is low. Spawning force_refresh_token()
```

I think it needs timeout for token retrieving.

**Comment by sigaloid** [comment]
Could be an issue with the atomic/concurrent nature of the token retrieval. Would appreciate if you could test after https://github.com/redlib-org/redlib/pull/158 to see if it has any difference

**Comment by freedit-dev** [comment]
Wow! Quick fix.

Deployed.

**Comment by freedit-dev** [comment]
@sigaloid hang again

```log
Jun 27 02:39:12 base redlib[4049503]:  TRACE redlib::client > Ratelimit remaining: 11.0
Jun 27 02:39:12 base redlib[4049503]:  TRACE redlib::client > Ratelimit used: 89
Jun 27 02:39:12 base redlib[4049503]:  TRACE redlib::client > Ratelimit reset: 48
Jun 27 02:39:14 base redlib[4049503]:  TRACE redlib::client > Ratelimit remaining: 10.0
Jun 27 02:39:14 base redlib[4049503]:  TRACE redlib::client > Ratelimit used: 90
Jun 27 02:39:14 base redlib[4049503]:  TRACE redlib::client > Ratelimit reset: 45
Jun 27 02:39:18 base redlib[4049503]:  TRACE redlib::client > Ratelimit remaining: 9.0
Jun 27 02:39:18 base redlib[4049503]:  TRACE redlib::client > Ratelimit used: 91
Jun 27 02:39:18 base redlib[4049503]:  TRACE redlib::client > Ratelimit reset: 41
Jun 27 02:39:18 base redlib[4049503]:  WARN  redlib::client > Rate limit 9 is low. Spawning force_refresh_token()
Jun 27 02:39:18 base redlib[4049503]:  TRACE redlib::oauth  > Rolling over refresh token. Current rate limit: 99
```

**Comment by sigaloid** [comment]
Hm. Okay, I see.

![image](https://github.com/redlib-org/redlib/assets/69441971/a95af9c4-0b4b-4fe0-95d5-2d6ca349ff05)
Side note, really funny that cloudflare's message here uses Redlib's CSS styling.

I'm going to work on trying to reproduce this locally first, some kind of load tester.

**Comment by freedit-dev** [comment]
Maybe you should not use async rwlock, using std rwlock instead.

**Comment by sigaloid** [comment]
I don't think it's a synchronization issue. It syncs on mine when I simulate high load.

I don't know what exactly the problem is - you're saying it's hanging precisely after the line "TRACE redlib::oauth  > Rolling over refresh token. Current rate limit: 99"? Nothing else?

**Comment by sigaloid** [comment]
here's an example of my log output: 

```
 TRACE redlib::client > Ratelimit remaining: 9.0
 TRACE redlib::client > Ratelimit used: 91
 TRACE redlib::client > Ratelimit reset: 103
 WARN  redlib::client > Rate limit 9 is low. Spawning force_refresh_token()
 TRACE redlib::oauth  > Rolling over refresh token. Current rate limit: 99
 INFO  redlib::oauth  > [✅] Success - Retrieved token "eyJhbGciOiJSUzI1NiIsImtpZCI6IlNI...", expires in 86399
 INFO  redlib::oauth  > Refreshing OAuth token... success
 TRACE redlib::client > Ratelimit remaining: 8.0
 TRACE redlib::client > Ratelimit used: 92
 TRACE redlib::client > Ratelimit reset: 102
 TRACE redlib::client > Ratelimit remaining: 5.0
 TRACE redlib::client > Ratelimit used: 95
 TRACE redlib::client > Ratelimit reset: 102
 TRACE redlib::client > Ratelimit remaining: 7.0
 TRACE redlib::client > Ratelimit used: 93
 TRACE redlib::client > Ratelimit reset: 102
 TRACE redlib::client > Ratelimit remaining: 6.0
 TRACE redlib::client > Ratelimit used: 94
 TRACE redlib::client > Ratelimit reset: 102
 WARN  redlib::client > Rate limit 6 is low. Spawning force_refresh_token()
 TRACE redlib::oauth  > Rolling over refresh token. Current rate limit: 99
 INFO  redlib::oauth  > [✅] Success - Retrieved token "eyJhbGciOiJSUzI1NiIsImtpZCI6IlNI...", expires in 86399
 INFO  redlib::oauth  > Refreshing OAuth token... success
 TRACE redlib::client > Ratelimit remaining: 3.0
```

Besides the occasional fact that a response may take longer to retrieve and the really small rate limits right before a refresh might accidentally overwrite the rate limit, my refreshing works just fine.

**Comment by freedit-dev** [comment]
Yes, it's occasional. And it hang exactly  `TRACE redlib::oauth > Rolling over refresh token. Current rate limit: 99`, nothing else. And restarting always fixed it.

```log
Jun 27 03:36:31 base redlib[4054219]:  WARN  redlib::client > Rate limit 8 is low. Spawning force_refresh_token()
Jun 27 03:36:31 base redlib[4054219]:  TRACE redlib::oauth  > Rolling over refresh token. Current rate limit: 99
Jun 27 03:36:31 base redlib[4054219]:  INFO  redlib::oauth  > [✅] Success - Retrieved token "eyJhbGciOiJSUzI1NiIsImtpZCI6IlNI...", expires in 86399
Jun 27 03:36:31 base redlib[4054219]:  INFO  redlib::oauth  > Refreshing OAuth token... success
Jun 27 03:36:31 base redlib[4054219]:  TRACE redlib::client > Ratelimit remaining: 7.0
Jun 27 03:36:31 base redlib[4054219]:  TRACE redlib::client > Ratelimit used: 93
Jun 27 03:36:31 base redlib[4054219]:  TRACE redlib::client > Ratelimit reset: 208
Jun 27 03:36:34 base redlib[4054219]:  WARN  redlib::client > Rate limit 7 is low. Spawning force_refresh_token()
Jun 27 03:36:34 base redlib[4054219]:  TRACE redlib::oauth  > Rolling over refresh token. Current rate limit: 99
Jun 27 03:36:34 base redlib[4054219]:  INFO  redlib::oauth  > [✅] Success - Retrieved token "eyJhbGciOiJSUzI1NiIsImtpZCI6IlNI...", expires in 86399
Jun 27 03:36:34 base redlib[4054219]:  INFO  redlib::oauth  > Refreshing OAuth token... success
Jun 27 03:36:34 base redlib[4054219]:  TRACE redlib::client > Ratelimit remaining: 99.0
Jun 27 03:36:34 base redlib[4054219]:  TRACE redlib::client > Ratelimit used: 1
Jun 27 03:36:34 base redlib[4054219]:  TRACE redlib::client > Ratelimit reset: 205
Jun 27 03:36:39 base redlib[4054219]:  TRACE redlib::client > Ratelimit remaining: 99.0
```

I just wanted to say that maybe we should have better logs to diagnose whether it is caused by "async rwlock" or "token retriving timeout".

**Comment by freedit-dev** [comment]
I just add more log and it stuck here

https://github.com/redlib-org/redlib/blob/3bd8b511a7234aabe045761ddc423a914608910a/src/oauth.rs#L74

And restarting fixed it.

**Comment by freedit-dev** [comment]
I'll try with a timeout.

```rust
trace!("Sending OAuth token request...");
let work = client.request(request);
let resp = match tokio::time::timeout(Duration::from_secs(10), work).await {
	Ok(result) => result.ok()?,
	Err(_) => {
		error!("Timeout: no response in 10 seconds.");
		return None;
	}
};
trace!("Received response from OAuth token request...");
```

**Comment by freedit-dev** [comment]
It's like some race condition

```log
Jun 27 15:58:43 base redlib[4141438]:  TRACE redlib::client > Ratelimit remaining: 10.0
Jun 27 15:58:43 base redlib[4141438]:  TRACE redlib::client > Ratelimit used: 90
Jun 27 15:58:43 base redlib[4141438]:  TRACE redlib::client > Ratelimit reset: 77
Jun 27 15:58:43 base redlib[4141438]:  TRACE redlib::client > Ratelimit remaining: 7.0
Jun 27 15:58:43 base redlib[4141438]:  TRACE redlib::client > Ratelimit used: 93
Jun 27 15:58:43 base redlib[4141438]:  TRACE redlib::client > Ratelimit reset: 76
Jun 27 15:58:43 base redlib[4141438]:  WARN  redlib::client > Rate limit 7 is low. Spawning force_refresh_token()
Jun 27 15:58:43 base redlib[4141438]:  TRACE redlib::oauth  > Rolling over refresh token. Current rate limit: 99
Jun 27 15:58:43 base redlib[4141438]:  TRACE redlib::oauth  > Trying to refresh OAuth token...
Jun 27 15:58:43 base redlib[4141438]:  TRACE redlib::oauth  > Sending OAuth token request...
Jun 27 15:58:43 base redlib[4141438]:  TRACE redlib::client > Ratelimit remaining: 9.0
Jun 27 15:58:43 base redlib[4141438]:  TRACE redlib::client > Ratelimit used: 91
Jun 27 15:58:43 base redlib[4141438]:  TRACE redlib::client > Ratelimit reset: 77
Jun 27 15:58:44 base redlib[4141438]:  WARN  redlib::client > Rate limit 9 is low. Spawning force_refresh_token()
```

two `force_refresh_token()` happened.

**Comment by freedit-dev** [comment]
> here's an example of my log output:
> 
> ```
>  TRACE redlib::client > Ratelimit remaining: 9.0
>  TRACE redlib::client > Ratelimit used: 91
>  TRACE redlib::client > Ratelimit reset: 103
>  WARN  redlib::client > Rate limit 9 is low. Spawning force_refresh_token()
>  TRACE redlib::oauth  > Rolling over refresh token. Current rate limit: 99
>  INFO  redlib::oauth  > [✅] Success - Retrieved token "eyJhbGciOiJSUzI1NiIsImtpZCI6IlNI...", expires in 86399
>  INFO  redlib::oauth  > Refreshing OAuth token... success
>  TRACE redlib::client > Ratelimit remaining: 8.0
>  TRACE redlib::client > Ratelimit used: 92
>  TRACE redlib::client > Ratelimit reset: 102
>  TRACE redlib::client > Ratelimit remaining: 5.0
>  TRACE redlib::client > Ratelimit used: 95
>  TRACE redlib::client > Ratelimit reset: 102
>  TRACE redlib::client > Ratelimit remaining: 7.0
>  TRACE redlib::client > Ratelimit used: 93
>  TRACE redlib::client > Ratelimit reset: 102
>  TRACE redlib::client > Ratelimit remaining: 6.0
>  TRACE redlib::client > Ratelimit used: 94
>  TRACE redlib::client > Ratelimit reset: 102
>  WARN  redlib::client > Rate limit 6 is low. Spawning force_refresh_token()
>  TRACE redlib::oauth  > Rolling over refresh token. Current rate limit: 99
>  INFO  redlib::oauth  > [✅] Success - Retrieved token "eyJhbGciOiJSUzI1NiIsImtpZCI6IlNI...", expires in 86399
>  INFO  redlib::oauth  > Refreshing OAuth token... success
>  TRACE redlib::client > Ratelimit remaining: 3.0
> ```
> 
> Besides the occasional fact that a response may take longer to retrieve and the really small rate limits right before a refresh might accidentally overwrite the rate limit, my refreshing works just fine.

Even success, there are two refresh at the same time.

**Comment by sigaloid** [comment]
That shouldn't be a problem, though. Independent threads can of course reach reddit simultaneously, I don't know why it would block in this case. There's nothing wrong with overwriting the token with a new one twice in succession, and the `RwLock` will stop any simultaneous modification.

Pushing a PR that should avoid the simultaneous token case, but I'm unsure if this will fix the problem. It rolls over on my machine under heavy load, but then again, so does `main` right now...

**Comment by sigaloid** [comment]
@freedit-dev @ButteredCats I know you both experienced this issue, please let me know if #160 helps. `latest` at quay.io has the newest version

**Comment by freedit-dev** [comment]
```log
Jun 28 03:49:51 base redlib[25514]:  TRACE redlib::client > Ratelimit remaining: Header says 62.0, we have 62.
Jun 28 03:49:51 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 38
Jun 28 03:49:51 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 8
Jun 28 03:49:53 base redlib[25514]:  TRACE redlib::client > Ratelimit remaining: Header says 61.0, we have 61.
Jun 28 03:49:53 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 39
Jun 28 03:49:53 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 6
Jun 28 03:49:53 base redlib[25514]:  TRACE redlib::client > Ratelimit remaining: Header says 60.0, we have 60.
Jun 28 03:49:53 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 40
Jun 28 03:49:53 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 6
Jun 28 03:49:55 base redlib[25514]:  ERROR redlib::utils  > Error page rendered: Nothing here
Jun 28 03:50:03 base redlib[25514]:  TRACE redlib::client > Ratelimit remaining: Header says 99.0, we have 59.
Jun 28 03:50:03 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 1
Jun 28 03:50:03 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 597
Jun 28 03:50:05 base redlib[25514]:  TRACE redlib::client > Ratelimit remaining: Header says 98.0, we have 58.
Jun 28 03:50:05 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 2
Jun 28 03:50:05 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 595
Jun 28 03:50:06 base redlib[25514]:  TRACE redlib::client > Ratelimit remaining: Header says 97.0, we have 57.
Jun 28 03:50:06 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 3
Jun 28 03:50:06 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 594
Jun 28 03:50:07 base redlib[25514]:  TRACE redlib::client > Ratelimit remaining: Header says 96.0, we have 56.
Jun 28 03:50:07 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 4
Jun 28 03:50:07 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 593
Jun 28 03:50:08 base redlib[25514]:  TRACE redlib::client > Ratelimit remaining: Header says 95.0, we have 55.
Jun 28 03:50:08 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 5
Jun 28 03:50:08 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 591
Jun 28 03:50:08 base redlib[25514]:  TRACE redlib::client > Ratelimit remaining: Header says 94.0, we have 54.
Jun 28 03:50:08 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 6
Jun 28 03:50:08 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 591
Jun 28 03:50:09 base redlib[25514]:  TRACE redlib::client > Ratelimit remaining: Header says 93.0, we have 53.
Jun 28 03:50:09 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 7
Jun 28 03:50:09 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 590
Jun 28 03:50:10 base redlib[25514]:  TRACE redlib::client > Ratelimit remaining: Header says 92.0, we have 52.
Jun 28 03:50:10 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 8
Jun 28 03:50:10 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 589
```

**Comment by freedit-dev** [comment]
@sigaloid So far so good. But I want to say `async rwlock` is not a good idea. It is not a frequent request and the current solution(async rwlock) is too complex.

From tokio doc https://docs.rs/tokio/latest/tokio/sync/struct.RwLock.html#method.read

> Note that under the priority policy of [RwLock](https://docs.rs/tokio/latest/tokio/sync/struct.RwLock.html), read locks are not granted until prior write locks, to prevent starvation. Therefore deadlock may occur if a read lock is held by the current task, a write lock attempt is made, and then a subsequent read lock attempt is made by the current task.

**Comment by sigaloid** [comment]
Hm. There's something I could possibly do to fix that - create a new Oauth struct and only call write to do an atomic swap, so the write call is held for far less time (just the time to do the swap, not the entire refresh call). Is that the kind of thing you mean?

Otherwise I'm not sure of a better solution besides a resource pool and that's a lot of overhead IMO vs just fixing the in place solution until it's performant

**Comment by freedit-dev** [comment]
`Async lock` is not an asynchronous version of std rwlock and is much more complex to deal with. Maybe you could try Arc_swap instead.

**Comment by freedit-dev** [comment]
#160 not work

```log
Jun 28 05:03:44 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 89
Jun 28 05:03:44 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 375
Jun 28 05:03:45 base redlib[25514]:  TRACE redlib::client > Ratelimit remaining: Header says 10.0, we have 10.
Jun 28 05:03:45 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 90
Jun 28 05:03:45 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 374
Jun 28 05:03:47 base redlib[25514]:  WARN  redlib::client > Rate limit 9 is low. Spawning force_refresh_token()
Jun 28 05:03:47 base redlib[25514]:  TRACE redlib::oauth  > Rolling over refresh token. Current rate limit: 8
Jun 28 05:03:47 base redlib[25514]:  TRACE redlib::client > Ratelimit remaining: Header says 9.0, we have 9.
Jun 28 05:03:47 base redlib[25514]:  TRACE redlib::client > Ratelimit used: 91
Jun 28 05:03:47 base redlib[25514]:  TRACE redlib::client > Ratelimit reset: 372
```

**Comment by sigaloid** [comment]
It doesn't roll over the tokens?

The logging order doesn't really matter, that's not indicative of an issue necessarily. but you're saying it hangs again?

**Comment by freedit-dev** [comment]
@sigaloid right, it hang again even for the latest commint.

Maybe we should see these again

> I just add more log and it stuck here
> 
> https://github.com/redlib-org/redlib/blob/3bd8b511a7234aabe045761ddc423a914608910a/src/oauth.rs#L74
> 
> And restarting fixed it.



> I'll try with a timeout.
> 
> ```rust
> trace!("Sending OAuth token request...");
> let work = client.request(request);
> let resp = match tokio::time::timeout(Duration::from_secs(10), work).await {
> 	Ok(result) => result.ok()?,
> 	Err(_) => {
> 		error!("Timeout: no response in 10 seconds.");
> 		return None;
> 	}
> };
> trace!("Received response from OAuth token request...");
> ```

Even timeout doesn't work.

**Comment by ButteredCats** [comment]
I updated and mine hung again within a minute, didn't seem to last any longer like freedit's.

When I can I'll enable the extra debugging and see if I have any logs to add, but that won't be until tonight and I doubt there's anything different from what freedit has sent.

**Comment by freedit-dev** [comment]
Restarting always fixed. So my instance auto restarts every 20mins.

**Comment by sigaloid** [comment]
Yeah, the issue is that the rollover doesn't work once the rate limit is hit. But the new token generation works at startup.


@freedit-dev, you mentioned arc-swap. I should have time tonight to push out a commit that switches to that. I assumed RwLock for the oauth client would be the best solution but honestly I'm not sure it is anymore. I don't know why it's not working, but I seriously doubt it's because of performance. CPU-based atomic are plenty to handle this workload of a few dozen requests a minute. On that note, what is your CPU architecture?

**Comment by freedit-dev** [comment]
```log
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         40 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  1
  On-line CPU(s) list:   0
Vendor ID:               GenuineIntel
  Model name:            Intel Core Processor (Broadwell, no TSX, IBRS)
    CPU family:          6
    Model:               61
    Thread(s) per core:  1
    Core(s) per socket:  1
    Socket(s):           1
    Stepping:            2
    BogoMIPS:            4788.90
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 syscall nx rdtscp lm constant_tsc rep_good nopl xtopology cpuid tsc_known_freq pni pclmu
                         lqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm cpuid_fault invpcid_single pti ssbd ibrs ibpb fsgsbase bmi1
                          avx2 smep bmi2 erms invpcid xsaveopt arat
Virtualization features: 
  Hypervisor vendor:     Microsoft
  Virtualization type:   full
Caches (sum of all):     
  L1d:                   32 KiB (1 instance)
  L1i:                   32 KiB (1 instance)
  L2:                    4 MiB (1 instance)
  L3:                    16 MiB (1 instance)
NUMA:                    
  NUMA node(s):          1
  NUMA node0 CPU(s):     0
Vulnerabilities:         
  Gather data sampling:  Not affected
  Itlb multihit:         KVM: Mitigation: VMX unsupported
  L1tf:                  Mitigation; PTE Inversion
  Mds:                   Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
  Meltdown:              Mitigation; PTI
  Mmio stale data:       Unknown: No mitigations
  Retbleed:              Not affected
  Spec rstack overflow:  Not affected
  Spec store bypass:     Mitigation; Speculative Store Bypass disabled via prctl
  Spectre v1:            Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:            Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
  Srbds:                 Unknown: Dependent on hypervisor status
  Tsx async abort:       Not affected
```

**Comment by sigaloid** [comment]
Interesting. Check out #162, let me know here if it helps

**Comment by ButteredCats** [comment]
Seems to have fixed it for me!

**Comment by sigaloid** [comment]
Splendid to hear!!

**Comment by freedit-dev** [comment]
So far so good. Thanks!

**Comment by sigaloid** [comment]
Amazing. Glad that I finally was able to fix it on the third go-around, huge thanks to @freedit-dev  for recommending the ArcSwap. I bet it was exactly the deadlock you mentioned. :thinking: I did limit scope to as small as possible but ArcSwap is the better long-term solution. Thanks @ButteredCats and @freedit-dev again for helping me debug this, as it really only showed up under heavy load! Closing :)
